convert/lib/convert.py

```diff
@@ -617,32 +690,23 @@ def delete_stack(self):
             ]
         }
 
-        def get_stack_resources(s) -> dict:
-            child_id = s["PhysicalResourceId"]
-            child_name = re.search(r":stack/(.*)/", child_id).group(1)
-            try:
-                if self.cf.describe_stacks(StackName=child_id)["Stacks"][0]["StackStatus"] == "DELETE_COMPLETE":
-                    return
-            except self.cf.exceptions.ClientError as e:
-                if "does not exist" in e.response["Error"]["Message"]:
-                    return
-                raise
-
-            child_resources = self.cf.describe_stack_resources(StackName=child_name)["StackResources"]
-            stacks[child_name] = [
-                r["LogicalResourceId"] for r in child_resources if r["ResourceStatus"] != "DELETE_COMPLETE"
-            ]
-            for r in child_resources:
-                if r["ResourceType"] == cdk_ids.cloudformation_stack.value:
-                    get_stack_resources(r)
-
         for s in root_resources:
             if s["ResourceType"] == cdk_ids.cloudformation_stack.value:
-                get_stack_resources(s)
+                self._get_stack_resources(s, stacks)
 
         for i, (stack, resources) in enumerate(stacks.items()):
             if self.args.delete:
-                self.cf.delete_stack(StackName=stack, RoleARN=cf_only_role["value"], RetainResources=resources)
+                if (
+                    stack_status := self.cf.describe_stacks(StackName=stack)["Stacks"][0]["StackStatus"]
+                ) and stack_status != "DELETE_FAILED":
+                    resources = None
```

Comment:
I'm still not 100% clear on what was actually going on with the CI failure (https://github.com/dominodatalab/cdk-cf-eks/actions/runs/5954998517/job/16152692447), and don't understand how the subsequent run with this code didn't exit with some failure. Note that it did leave all the nested stacks behind, so wasn't a true success, though I unfortunately destroyed the evidence as I cleaned them up to test that this stack deletion strategy still works in general).

Anyway, it sounds almost as if we got to this line of code without the stacks actually having a deletion attempt. We should not be able to get this point without the stacks being in `DELETE_FAILED`. The whole point of this workflow is that we intentionally delete the stacks in a way where they will all fail, which triggers the state where we can provide the list of resources. Makes me wonder if AWS changed it so the child stacks aren't ending up in `DELETE_FAILED`, and what we really need to do is iterate when we run it the first time.

We should probably also iterate and confirm that none of them are in non-`DELETE_FAILED` states, and report what those states are and exit with a failure, before we get to that second delete. Anything other than `DELETE_FAILED` (or `DELETE_COMPLETE` if you check before the line where we take those out) indicates that something went wrong.

Reply:
Initially misunderstood scope, Fixed by iterating through nested stacks